model/qti/asset/handler/StimulusHandler.php

```diff
@@ -0,0 +1,188 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2018 (original work) Open Assessment Technologies SA;
+ *
+ */
+
+namespace oat\taoQtiItem\model\qti\asset\handler;
+
+use oat\taoItems\model\media\LocalItemSource;
+use oat\taoQtiItem\model\qti\Element;
+use oat\taoQtiItem\model\qti\Item;
+
+/**
+ * Class StimulusHandler
+ *
+ * The purpose of this handler is to extract xml stimulus from item to copy them
+ * If the included stimulus includes images, they should be encoded to avoid file dependencies
+ *
+ * @package oat\taoQtiItem\model\qti\asset\handler
+ */
+class StimulusHandler implements AssetHandler
+{
+    /**
+     * The object representation of qti item
+     *
+     * @var Item
+     */
+    protected $qtiItem;
+
+    /**
+     * The destination of the item directory
+     *
+     * @var LocalItemSource
+     */
+    protected $itemSource;
+
+    /**
+     * Applicable to all items that contain <xinclude> tag
+     *
+     * @param $relativePath
+     * @return bool
+     * @throws \common_exception_MissingParameter
+     */
+    public function isApplicable($relativePath)
+    {
+        $xincluded = array();
+        /** @var Element $xincludeElement */
+        foreach ($this->getQtiItem()->getComposingElements('oat\taoQtiItem\model\qti\Xinclude') as $xincludeElement) {
+            $xincluded[] = $xincludeElement->attr('href');
+            \common_Logger::i("Xinclude component found in resource '" .
+                $this->getQtiItem()->getIdentifier() . "' with href '" . $xincludeElement->attr('href') . "'.");
+        }
+
+        return in_array($relativePath, $xincluded);
+    }
+
+    /**
+     * Store locally, in a safe directory
+     * Encoded all stimulus images to avoid file dependencies
+     *
+     * @param $absolutePath
+     * @param $relativePath
+     * @return array|mixed
+     * @throws \common_Exception
+     */
+    public function handle($absolutePath, $relativePath)
+    {
+        $safePath = $this->safePath($relativePath);
+        $this->encodeStimulusImages($absolutePath);
+
+        $info = $this->getItemSource()->add($absolutePath, basename($absolutePath), $safePath);
+        \common_Logger::i('Stimulus file \'' . $absolutePath . '\' copied.');
+
+        return $info;
+    }
+
+    /**
+     * Remove ../ to secure path
+     *
+     * @param $path
+     * @return string
+     */
+    protected function safePath($path)
+    {
+        $safePath = '';
+        if (dirname($path) !== '.') {
+            $safePath = str_replace('../', '', dirname($path)) . '/';
```

Comment:
if str_replace find `../` in the middle of the `$path` it will replace it too

Reply:
yes to avoid exploit like `toto/../../../../../other/server/file`. It will created a directory with valid part and put it under item directory